src/shader_recompiler/frontend/translate/vector_memory.cpp

```diff
@@ -435,6 +439,46 @@ void Translator::BUFFER_STORE_FORMAT(u32 num_dwords, bool is_typed, bool is_form
     }
 }
 
+void Translator::BUFFER_ATOMIC(u32 num_dwords, AtomicOp op, const GcnInst& inst) {
+    // Get controls for mubuf-specific instructions
+    const auto& mubuf = inst.control.mubuf;
+    // Get vaddr register
+    const IR::VectorReg vaddr{inst.src[0].code};
+    // Get vdata register
+    const IR::VectorReg vdata{inst.src[1].code};
+    // Get srsrc register
+    const IR::ScalarReg srsrc{inst.src[2].code * 4};
+    // Get offset value from soffset register
+    const IR::U32 soffset{GetSrc(inst.src[3])}; // TODO: Use this maybe?
+    ASSERT_MSG(soffset.IsImmediate() && soffset.U32() == 0, "Non immediate offset not supported");
+
+    // Setup instruction info from controls
+    IR::BufferInstInfo info{};
+    info.index_enable.Assign(mubuf.idxen);
+    info.inst_offset.Assign(mubuf.offset);
+    info.offset_enable.Assign(mubuf.offen);
+
+    // Get vdata value
+    IR::Value vdata_val = ir.GetVectorReg<Shader::IR::U32>(vdata);
+
+    // Get address of vdata
+    const IR::U32 address = ir.GetVectorReg(vaddr);
+
+    // Construct srsrc SGPRs
+    const IR::Value handle =
+        ir.CompositeConstruct(ir.GetScalarReg(srsrc), ir.GetScalarReg(srsrc + 1),
+                              ir.GetScalarReg(srsrc + 2), ir.GetScalarReg(srsrc + 3));
+
+    // Apply atomic op
+    // derefs srsrc buffer and adds vdata value to it
+    // then returns original srsrc buffer value
+    const IR::Value original_val = ir.BufferAtomicIAdd(handle, address, vdata_val, info);
```

Comment:
I think the switch that was here is fine, since most buffer atomics are implemented on backend side anyway.